pkg/controllers/user/globaldns/user_cluster_globaldns_controller.go

```diff
@@ -56,67 +77,130 @@ func (g *UserGlobalDNSController) sync(key string, obj *v3.GlobalDNS) (runtime.O
+			return nil, fmt.Errorf("UserGlobalDNSController: Error in splitting project Name %v", projectNameSet)
+		}
+		// check if the project in this iteration belongs to the same cluster in current context
+		if split[0] != g.clusterName {
+			logrus.Debugf("UserGlobalDNSController: Continuing since project %v doesn't belong in cluster %v", split[1], g.clusterName)
+			continue
+		}
+		projectID := split[1]
+		//list all namespaces in this project and list all ingresses within each namespace
+		namespaces, err := g.namespaceLister.List("", labels.NewSelector())
+		if err != nil {
+			return nil, fmt.Errorf("UserGlobalDNSController: Error listing cluster namespaces")
+		}
+		var namespacesInProject []*corev1.Namespace
+		for _, namespace := range namespaces {
+			nameSpaceProject := namespace.ObjectMeta.Labels[projectSelectorLabel]
+			if strings.EqualFold(projectID, nameSpaceProject) {
+				namespacesInProject = append(namespacesInProject, namespace)
+			}
+		}
+		for _, namespace := range namespacesInProject {
+			ingresses, err := g.ingressLister.List(namespace.Name, labels.NewSelector())
+			if err != nil {
+				return nil, err
+			}
+			allIngresses = append(allIngresses, ingresses...)
+		}
 	}
+	//gather endpoints
+	return g.fetchGlobalDNSEndpointsForIngresses(allIngresses, obj)
+}
+
+func (g *UserGlobalDNSController) fetchGlobalDNSEndpointsForIngresses(ingresses []*v1beta1.Ingress, obj *v3.GlobalDNS) ([]string, error) {
 	if len(ingresses) == 0 {
 		return nil, nil
 	}
 
+	var allEndpoints []string
+	//gather endpoints from all ingresses
 	for _, ing := range ingresses {
 		if gdns, ok := ing.Annotations[annotationGlobalDNS]; ok {
 			// check if the globalDNS in annotation is same as the FQDN set on the GlobalDNS
 			if gdns != obj.Spec.FQDN {
 				continue
 			}
-			//update endpoints on GlobalDNS status field
+			//gather endpoints from the ingress
 			ingressEndpoints := gatherIngressEndpoints(ing.Status.LoadBalancer.Ingress)
-			updatedGDNS, err := g.updateGlobalDNSEndpoints(obj, ingressEndpoints)
-			if err != nil {
-				return updatedGDNS, err
-			}
+			allEndpoints = append(allEndpoints, ingressEndpoints...)
 		}
 	}
-	return nil, nil
+
+	return allEndpoints, nil
 }
 
-func (g *UserGlobalDNSController) updateGlobalDNSEndpoints(globalDNS *v3.GlobalDNS, ingressEndpoints []string) (*v3.GlobalDNS, error) {
-	globalDNS = prepareGlobalDNSForUpdate(globalDNS, ingressEndpoints)
-	updated, err := g.globalDNSs.Update(globalDNS)
+func (g *UserGlobalDNSController) refreshGlobalDNSEndpoints(globalDNS *v3.GlobalDNS, ingressEndpointsForCluster []string) (*v3.GlobalDNS, error) {
+
+	mapIngressEndpoints := make(map[string]bool)
+	for _, ep := range ingressEndpointsForCluster {
+		mapIngressEndpoints[ep] = true
+	}
+	//find endpoints removed by looking at clusterEndpoints of current cluster
+	endpointsToRemove := []string{}
+
+	if len(globalDNS.Status.ClusterEndpoints) == 0 {
+		globalDNS.Status.ClusterEndpoints = make(map[string][]string)
```

Comment:
looks like we are updating the object here w/o doing the deepCopy

Reply:
edited